### PR TITLE
Fix padding gradients in vocab-parallel embeddings

### PR DESCRIPTION
**Pull request by zupengwang**
The HF backend preserves `nn.Embedding.padding_idx` when converting embeddings to TorchTitan modules. Vocab-parallel forward then passes that global index to a local weight shard. For SmolLM3's `vocab_size=128256` and `pad_token_id=128004`, TP=2 produces 64128-row local tables and the old forward raises `Padding_idx must be within num_embeddings`. Lower padding IDs can silently suppress gradients for unrelated rows on other shards.

Map the padding index to local coordinates only on its owning shard. Expose `padding_idx: int | None = None` in `Embedding.Config` and pass it to the PyTorch constructor, so padding support is explicit for native modules as well as HF-converted embeddings. Existing model configs retain the default `None` behavior.

The CPU tests cover native config construction, negative indices, initialization, checkpoint loading with nonzero padding weights, and forward/backward parity. The distributed HF conversion regression covers 24 combinations of vocabulary size, padding index, and token coverage. The FSDP=2/TP=2 regression uses native `Config.build()` and covers ten padding/reshard combinations over three optimizer steps, checking parameter restoration, gradients, and state_dict parity.

Validation of the source committed as `1cbdfc198aba6287fe4611b2ea246bde6e6dde4a`, on remote RTX 3090 GPUs with PyTorch 2.15.0.dev20260911+cu130 and Transformers 5.9.0:

- 25 embedding/linear CPU tests and 11 embedding/FSDP tests on four GPUs passed.
- The new config regression fails on the previous head because `padding_idx` is not an accepted configuration field.
- Native Llama3 and HF-backend FSDP=2/TP=2 training each completed 10 steps before and after the config change. All avg/max losses and grad_norm values matched bitwise at TensorBoard float32 precision with seed 42 and strict deterministic mode.
- The training runs used randomly initialized debug models, sequence length 128, 512 tokens/step, and local C4 data. The HF model retained the full 128256 vocabulary and actual padding ID 128004. Native loss was 8.22354 -> 4.39290; HF loss was 12.08252 -> 8.71970.
- Scoped pre-commit and `git diff --check` passed. Whole-tree pre-commit reproduces the existing formatting issue in `torchtitan/experiments/rl/tests/test_dapo_math.py`; that file is unchanged.

The earlier frequency-scaling/max_norm support was removed in response to review. Those non-default fields retain main's behavior; this PR does not claim to fix TP frequency scaling or FSDP2 max_norm persistence. These checks establish debug-model regression parity, not full-model quality or throughput improvements.

### CODE REVIEW

**Comment by tianyu-l** [comment]
comment 1: this model sounds a bit outdated

comment 2: if you rely on transformers, checkout the transformers_modeling_backend folder

**Comment by zupengwang** [comment]
Thanks for the feedback, @tianyu-l. I've removed the separate SmolLM3 experiment and narrowed this PR to a shared embedding issue found while testing `transformers_modeling_backend`.

Vocab-parallel embedding was using a global `padding_idx` against local weight shards. This can either raise an assertion or silently suppress gradients for unrelated rows. Commit `8769911` maps the index only on its owning shard and adds distributed forward/backward regression coverage.

Validation: 24 CPU tests and 9 GPU tests passed, along with 10-step HF-backend runs including the randomly initialized full 3B architecture with FSDP=2 / TP=2. Single-GPU loss and `grad_norm` are identical before and after the fix.

Would you prefer this fix here or in a separate PR?

File: torchtitan/models/common/embedding.py [hunk 1]
```diff
@@ -50,10 +50,17 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
...
+        if (
+            self.padding_idx is not None
+            and offset <= self.padding_idx < offset + self.weight.shape[0]
+        ):
+            local_padding_idx = self.padding_idx - offset
         out = F.embedding(
             local_input,
             self.weight,
-            self.padding_idx,
+            local_padding_idx,
```
Comment (mlazos) [inline torchtitan/models/common/embedding.py:66]: _Reviewed by me, this may be an issue._

_AI-generated analysis (Codex):_

> **Masked IDs distort frequency-scaled gradients**
> 
> With scale_grad_by_freq=True, F.embedding counts clamped nonlocal IDs before the output mask removes them. For vocab 128, TP=4, padding_idx=0, and tokens [0, 32], rank 1 receives local indices [0, 0] with local_padding_idx=None, so global row 32 gets gradient 1/2 instead of 1. The added test uses the default scale_grad_by_freq=False and does not cover this incomplete padding-gradient fix.

Comment (zupengwang) [reply torchtitan/models/common/embedding.py:66]: Thanks, confirmed. Your `[0, 32]` example produces a gradient of 0.5 instead of 1 on commit `8769911`. Fixed in `60095e3`.

The follow-up maps nonlocal tokens to a separate temporary row when `scale_grad_by_freq=True`, so they no longer inflate any real row's frequency. It also preserves `max_norm`'s in-place weight updates when using the padded table. The default path continues to use the original weight directly.

The expanded distributed regression covers both frequency modes, uneven vocabularies, padding boundaries, repeated tokens, ranks with no local hits, and renormalization. All 72 combinations match the full `nn.Embedding` outputs, gradients, and updated weights exactly on CPU and four GPUs. The CPU suite has 24 passing tests and the GPU suite has 9. TP=2 training with frequency scaling enabled completes 10 steps; the default path's loss and grad_norm match the previous head exactly for all 10 steps.

**Comment by tianyu-l** [comment]
@claude review this PR. In particular
- show a simple example where without the fix it's problematic
- does it work well with FSDP2's `FSDPModule` lifecycle?

**Comment by zupengwang** [comment]
Thanks, @tianyu-l. I added a four-GPU FSDP2/TP regression in `6a4ca3a` and checked the lifecycle caveat experimentally.

For a simple failure: with vocabulary 128 and TP=4, rank 1 owns global rows 32–63. With `padding_idx=5`, token 37 maps to local row 5, and the original code incorrectly suppresses its gradient: 0 instead of 1. A global padding ID of 40 instead raises `Padding_idx must be within num_embeddings`. Both examples now match `nn.Embedding`.

The new test uses `Module.parallelize` and nested `fully_shard` modules with FSDP=2/TP=2. It covers 16 combinations of padding, frequency scaling, and reshard-after-forward settings, each over three optimizer steps with different DP batches. Outputs, gradients, and saved weights match the reference exactly; it also checks parameter restoration for the optimizer. Substituting the earlier `8769911` forward makes this test fail on frequency-scaled gradients.

The `max_norm` caveat is real: both vanilla `nn.Embedding + FSDP2` and this TP implementation produce normalized outputs but discard the forward-time weight mutation after resharding, with frequency scaling on or off. I added a TODO documenting this existing limitation.

Local validation: 9 CPU tests and 10 tests on four GPUs passed. A 10-step HF-backend debug-model run with FSDP=2/TP=2 and frequency scaling enabled completed (loss 12.08252 -> 8.74190); loss and grad_norm matched the previous head bit-for-bit at TensorBoard precision. The [CPU CI log](https://github.com/pytorch/torchtitan/actions/runs/35040740171/job/104636056799) also confirms the existing padding sweep runs and passes via the regular CPU workflow, without the fake-PG label.

**Comment by tianyu-l** [comment]
@claude I'm not convinced that we need to support those fields.

In torchtitan, our Embedding module inherit nn.Embeddings, and are only passing `num_embeddings`, `embedding_dim` to the nn.Embedding's constructor below
```
class torch.nn.Embedding(num_embeddings, embedding_dim, padding_idx=None, max_norm=None, norm_type=2.0, scale_grad_by_freq=False, sparse=False, _weight=None, _freeze=False, device=None, dtype=None)
```

My impression is that we will NEVER hit the issues worried by this PR.

**Comment by tianyu-l** [comment]
@zupengwang do you care about the HF backend in torchtitan?

Only embedding is being shared here; all others are not sharing torchtitan module https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/transformers_modeling_backend/module_conversion.py#L56

I don't know if it's worth the exception.

**Comment by zupengwang** [comment]
Yes—the concrete use case is running SmolLM3 through `transformers_modeling_backend`, following your earlier pointer to that backend. I agree that native `Embedding.Config` instances never set these fields.

I traced the actual HF construction/conversion path: SmolLM3's config has `vocab_size=128256` and `pad_token_id=128004`. Its `nn.Embedding` retains that padding ID after `convert_hf_to_module`. At TP=2 the local table has 64128 rows. With the pre-PR forward, a backend training run using those real vocabulary/padding values fails on the first embedding lookup with `Padding_idx must be within num_embeddings`. I reduced hidden/layer sizes for the run; I did not inject embedding attributes.

I've trimmed the change in `e107b8b`: the temporary table, frequency-scaling handling, max_norm writeback, and associated TODO are removed. Those non-default options retain main's handling and are outside this PR's scope. The remaining production diff is 8 added lines and one replacement in the already-shared Embedding, mapping padding to the owning vocabulary shard.

The CPU regression now constructs an HF Llama model and calls the actual conversion function before checking TP outputs/gradients. The padding-only FSDP2/TP2 lifecycle test is retained. Validation passed: 24 CPU tests, 10 tests on four GPUs, and 10-step backend training before/after the trim with identical loss and grad_norm at TensorBoard precision.

File: torchtitan/models/common/embedding.py
Comment (tianyu-l) [inline torchtitan/models/common/embedding.py:29]: With HF, the usage of `self.padding_idx` is implicit -- how would the maintainer of this Embedding module know that someone is going to plug in a `self.padding_idx`?

Can we put `padding_idx` in the config and default to `None`, at least?

Comment (zupengwang) [reply torchtitan/models/common/embedding.py:29]: Added `padding_idx: int | None = None` to `Embedding.Config` in `1cbdfc1` and passed it to the `nn.Embedding` constructor. This makes padding an explicit module option and retains PyTorch's index validation and negative-index normalization. Existing configs continue to default to `None`.

The new CPU regression checks native config construction, initialization, checkpoint loading with nonzero padding weights, and forward/backward parity with `nn.Embedding`. The FSDP=2/TP=2 test now constructs the embedding through `Config.build()`, including padding and sharding options, and covers both reshard modes over three optimizer steps. The HF conversion regression is retained.

Remote validation passed: 25 CPU tests and 11 tests on four GPUs. Native Llama3 and HF-backend training each ran 10 steps before/after this change with seed 42 and strict deterministic mode; every loss and grad_norm matched bitwise at TensorBoard float32 precision. Scoped pre-commit also passed.

**Review by tianyu-l — APPROVED** [review]
